DSpace/DSpace
# Support IIIF manifest bundle on items

## References
* Related to https://github.com/DSpace/DSpace/issues/9143

## Description
Not mergeable, but looking for feedback, especially since this is the first time I'm poking at the DSpace codebase.

While the related issue initially talks about updating the IIIF manifest generation to be more configurable, it was suggested that supporting externally created IIIF manifest files on items would be a better option, allowing for more robust IIIF manifests without expanding the DSpace scope unnecessarily.

This does a minimum to serve an existing Bitstream from the IIIF_MANIFEST bundle on an item, if it exists, otherwise it will follow the current path of creating (and caching) a "default" IIIF manifest

## Instructions for Reviewers

List of changes in this PR:
* Add constant for new Bundle name
* Update ManifestService to serve bitstream `IIIF_MANIFEST` bundle if present
* Add ITs

There are still some TODOs if these changes are going in the right direction:
* The rest of the IIIF endpoints need to be modified to pull data from the IIIF_MANIFEST bundle content (see methods in the [IIIFServiceFacade](https://github.com/DSpace/DSpace/blob/main/dspace-iiif/src/main/java/org/dspace/app/iiif/IIIFServiceFacade.java))

### Testing

**Running ITs:**
`mvn install -DskipIntegrationTests=false -Dit.test=org.dspace.app.rest.iiif.IIIFControllerIT -DfailIfNoTests=false`

**If manual testing:**
* You'll need to rebuild the dspace-iiif component and the webapp, making sure configs are set to have IIIF enabled
* Start up the DSpace application, find an item to investigate
* Enable IIIF on the item by adding `dspace.iiif.enabled=true` to the item metadata and check its default IIIF manifest content: `<dspace-server>/server/iiif/<obj_id>/manifest`
* Add a new Bitstream under a new `IIIF_MANIFEST` bundle on the item using a custom file. At the moment, there are no checks for file types or anything so you can pretty much use anything, but there are sample JSON files (that may or may not pass IIIF validation checks) at `dspace-server-webapp/src/test/resources/org/dspace/app/rest/iiif/` and re-check the manifest to see that the returned content should have changed to the content of the file you just uploaded

## Some questions:

* Any validation on uploaded file? I'm not sure if DSpace has some conventions for file upload validation, or if that's left for application admins. Minimum it should have `application/json` mime-type, even if we don't want to do full IIIF manifest validation
* Any interest in have a dspace-cli operation to generate IIIF manifests for applicable items offline, similar to generating thumbnails (would be a separate PR if so)
* It would be nice to show `IIIF_MANIFEST` (or whatever name it would be) in the admin UI when selecting a bundle to add a bitstream to. Is the prepopulated list of bundle names a frontend or backend change? Ex: 
![image](https://github.com/DSpace/DSpace/assets/5167587/b3269d66-7c09-4a5c-b7ba-fb76029ca279)


## TODO:

- ~[ ] Update `searchInManifest`~
  - Presumably, if someone adds their own manifest that has search support, they'll point to the already existing DSpace IIIF search, or their own custom search server. DSpace IIIF search will be handled the same way, whereas with custom search, Mirador will send the request to that service, skipping the DSpace search endpoint
- [x] Update `getSeeAlsoAnnotations`
- [ ] Update `dspace-angular` to add IIIF_MANIFEST bundle when creating a new bundle on things with IIIF enabled


## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
